### ISSUE TITLE
clean datafile on segmented data with getSegmentLimits

### ISSUE THREAD

**Issue by zack-vii** [MDSplus/mdsplus#335]
We experience problem with segmented data (images, shape per segment (1,2048,2048), i.e. one frame per segment) the segments have been written in parallel to other similar data. assumption the segments stored in the data file are not one continuous  stream. If TCL command clean / python method cleanDatafile is used on the tree, the functionality of getSegmentLimits is broken. returning either None  or [Cmplx($ROPRAND,$ROPRAND),Cmplx($ROPRAND,$ROPRAND)]. It happens on all 20 CYGNET4K frames on all shots. I am currently working on a reproducible example with less data. I will update on the results.

**Comment by tfredian** [comment]
Hmmm. Couple questions:
1) Is this just a problem with getSegmentLimits or do the segments themselves get trashed by a clean?
2) Can you build using the alpha head and test that? I reworked the clean/compress code somewhat and added the ability to do: "TCL> COMPRESS treename/SHOT=shotnum" and compress signals. I did only limited testing on this feature and mostly on timestamped segments.

**Comment by zack-vii** [comment]
I figured out a small sample script that reproduces the conditions. It seems to be related to the compression of the segments. Obviously the Cleanup routine does not check for the compressions flag. It reads the compressed datastream as if it were uncompressed and thereby get mixed up with the datatypes (well everything). You can use the following script to generate a minimal sample tree with data that runs into the problem. without the compress_segments flag set the clean up does not cause any trouble.

``` python
"""clean datafile problem"""
from MDSplus import Tree, Int16Array,Float32Array
# create sample tree
with Tree('TEST',1,'new') as tree:
    tree.addNode('SEG','SIGNAL').compress_segments=True
    tree.write()
# open sample tree for writing sample data
seg = Tree('TEST',1).SEG
data = Int16Array([1])
for dim in range(10):
    seg.makeSegment(dim,dim,Float32Array([dim]),data)
# check before cleaning
print(seg.data().shape)
print(seg.getSegmentStart(3))
# removing references before cleaning up data file
del(seg)
Tree('TEST',1).cleanDatafile()
# reopen tree to check after cleaning
seg = Tree('TEST',1).SEG
print(seg.data().shape)
print(seg.getSegmentStart(3)) # the error occurs
```

**Comment by tfredian** [comment]
Thanks for the info. Are you testing this with the current alpha branch head or a particular MDSplus release?

**Comment by zack-vii** [comment]
Well the original issue occurred based on a 2217806 pull request #300 (version7-0-206 +1commit). I installed the newest alpha on a test machine (unix) and this time the clean does also corrupt the data itself. i.e. the line 

``` python
print(seg.data().shape)
```

after the cleanDatafile causes a segfault

**Comment by zack-vii** [comment]
gbd says:

```
Program received signal SIGSEGV, Segmentation fault.
XTreeDefaultResampleMode (startD=startD@entry=0x0, endD=endD@entry=0x0,
    inDeltaD=inDeltaD@entry=0x0, mode=mode@entry=1 '\001',
    outSignalXd=outSignalXd@entry=0xa2e888, inSignalD=0x7fffffff9d80,
    inSignalD=0x7fffffff9d80) at XTreeDefaultResample.c:605
605     XTreeDefaultResample.c: No such file or directory.
```

**Comment by zack-vii** [comment]
while seg.getSegment(0) works fine - i.e. the data is intact - reg.record fails. Most likely that the unification of the segments fails

**Comment by tfredian** [comment]
So I'm really confused now. I just built the alpha head on fc23 and ran your python test above and it worked fine!

[twf@fc23-vm test]$ python test.py
(10,)
3
(10,)
3

where test.py contains the python test statements above.

**Comment by zack-vii** [comment]
hm.. that is strange.. I  am not entirely sure about the version that we are running on the MDSplus server as we needed to tweak the compression fix #300 manually to get a running version before operation at the 10th of January. however, have you tried the most recent alpha build 7.0.213

**Comment by tfredian** [comment]
I just now tried it on build 7.0.213 and it also worked fine.

**Comment by zack-vii** [comment]
Ah I checkt again the git log. We manually applied 'refs/remotes/MDSplus/twf-fix-compress-again2' onto the version I mentioned. Unsure if this will explain anything. That is strange. cause I dis a clean install of alpha 213 just before trying. I will investigate further tomorrow

**Comment by zack-vii** [comment]
Hm, tested again on my Win10 64bit:
MDSplus-7.0-62 (stable) fails in line

``` python
print(seg.getSegmentStart(3)) # the error occurs
```

MDSplus-alpha-7.0-213 fails in line

``` python
print(seg.data().shape)
```

which is consistent with out unix results (Debian Weezel, 64bit).
I don't know much about the fc23. Is it possibly 32bit? Is 64bit the issue again?

**Comment by zack-vii** [comment]
Nope, its not the 64bit thingy, I set upo WinXp 32bit on a VM, and installed MDSplus-alpha-7.0-213.
same error in line 20

``` python
print(seg.data().shape)
```

 WindowsError: exception: access violation reading 0x00000002

**Comment by tfredian** [comment]
Ok, I ran the test again with valgrind and discovered that there was an invalid if statement on an status value when the status was not yet initialized. I put in a pull request  https://github.com/MDSplus/mdsplus/pull/337 to fix this problem. With this fix valgrind completes the test without any errors. If you can do so could you test this on your Wheezy system using the pull request branch?

**Comment by zack-vii** [comment]
indeed likely fixed by #337
